### DISCUSSION TITLE
error：undefined reference to sth

### DISCUSSION THREAD

**Discussion by Darkhell101** [alandefreitas/matplotplusplus#197] [Q&A]
I installed binary pkg on ubuntu 21.04 to put into my study.
And I use "g++ -std=c++17 test.cpp -o test" to run a file. Some errors happens. Please help me.
The error is like the following:
/tools$ g++ -std=c++17 test.cpp -o test
/usr/bin/ld: /tmp/ccae4fSO.o: in function `main':
test.cpp:(.text+0x289): undefined reference to `matplot::gcf()'
/usr/bin/ld: test.cpp:(.text+0x2b2): undefined reference to `matplot::figure_type::width() const'
/usr/bin/ld: test.cpp:(.text+0x2be): undefined reference to `matplot::figure_type::width(unsigned int)'
/usr/bin/ld: test.cpp:(.text+0x2e2): undefined reference to `matplot::subplot(unsigned long, unsigned long, unsigned long, bool)'
/usr/bin/ld: test.cpp:(.text+0x387): undefined reference to `matplot::title(std::basic_string_view<char, std::char_traits<char> >)'
/usr/bin/ld: test.cpp:(.text+0x3ab): undefined reference to `matplot::subplot(unsigned long, unsigned long, unsigned long, bool)'
/usr/bin/ld: test.cpp:(.text+0x455): undefined reference to `matplot::title(std::basic_string_view<char, std::char_traits<char> >)'
/usr/bin/ld: /tmp/ccae4fSO.o: in function `matplot::show()':
test.cpp:(.text._ZN7matplot4showEv[_ZN7matplot4showEv]+0x24): undefined reference to `matplot::gcf()'
/usr/bin/ld: test.cpp:(.text._ZN7matplot4showEv[_ZN7matplot4showEv]+0x38): undefined reference to `matplot::figure_type::show()'
/usr/bin/ld: /tmp/ccae4fSO.o: in function `matplot::save(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
test.cpp:(.text._ZN7matplot4saveERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZN7matplot4saveERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x28): undefined reference to `matplot::gcf()'
/usr/bin/ld: test.cpp:(.text._ZN7matplot4saveERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZN7matplot4saveERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x46): undefined reference to `matplot::figure_type::save(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: /tmp/ccae4fSO.o: in function `auto matplot::area<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >, double>(std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >, double&&)':
test.cpp:(.text._ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJdEEEDaT_DpOT0_[_ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJdEEEDaT_DpOT0_]+0x30): undefined reference to `matplot::gca()'
/usr/bin/ld: test.cpp:(.text._ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJdEEEDaT_DpOT0_[_ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJdEEEDaT_DpOT0_]+0x8c): undefined reference to `matplot::axes_type::area(std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > const&, double, bool, std::basic_string_view<char, std::char_traits<char> >)'
/usr/bin/ld: /tmp/ccae4fSO.o: in function `auto matplot::area<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >, int, bool>(std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >, int&&, bool&&)':
test.cpp:(.text._ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJibEEEDaT_DpOT0_[_ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJibEEEDaT_DpOT0_]+0x36): undefined reference to `matplot::gca()'
/usr/bin/ld: test.cpp:(.text._ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJibEEEDaT_DpOT0_[_ZN7matplot4areaISt6vectorIS1_IdSaIdEESaIS3_EEJibEEEDaT_DpOT0_]+0xad): undefined reference to `matplot::axes_type::area(std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > const&, double, bool, std::basic_string_view<char, std::char_traits<char> >)'
collect2: error: ld returned 1 exit status

**Comment by alandefreitas** [comment]
You also need to link the library in `g++ -std=c++17 test.cpp -o test -l <...>` 

Linking options: https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html

**Reply by Darkhell101** [reply]
still not work😔
